snapshotter/processor_distributor.py

```diff
@@ -978,6 +1049,24 @@ async def _rabbitmq_consumer(self, loop):
             await q_obj.bind(exchange, routing_key=self._consume_queue_routing_key)
             await q_obj.consume(self._on_rabbitmq_message)
 
+    async def _cleanup_tasks(self):
+        """
+        Periodically clean up completed or timed-out tasks.
+        """
+        while True:
+            await asyncio.sleep(self._task_cleanup_interval)
+            for task_start_time, task in list(self._active_tasks):
+                current_time = time.time()
+                if task.done():
+                    self._active_tasks.discard((task_start_time, task))
+
+                elif current_time - task_start_time > self._task_timeout:
+                    self._logger.warning(
+                        f'Task {task} timed out. Cancelling..., current_time: {current_time}, start_time: {task_start_time}',
+                    )
+                    task.cancel()
+                    self._active_tasks.discard((task_start_time, task))
```

Comment:
I believe it's best practice to await a cancelled task to ensure proper cleanup and avoid the event loop attempting to operate on a cancelled task. Something like the solution outlined here should cover it: https://github.com/python/cpython/issues/103486#issue-1665155187

I don't think this feature was ever added to the core library.

Reply:
Implemented in https://github.com/PowerLoom/snapshotter-core/pull/11/commits/1b39404000e0cf2fe5cbae73158cfc78e7dd8a93.

Reply:
![Screenshot 2024-10-02 at 2 45 35 PM](https://github.com/user-attachments/assets/ab55b996-fb26-46c2-8ded-50fc42417016)
@Seth-Schmidt, this change is causing too much memory and CPU usage; reverting for now. If needed, we can do this in a new PR.

Reply:
@xadahiya Gotcha. That works for me.